### DISCUSSION TITLE
Lookup/recovery codes stored in cleartext?

### DISCUSSION THREAD

**Discussion by kszafran** [ory/kratos#2676] [Q&A]
@aeneasr I just noticed that 2FA lookup/recovery codes are stored in cleartext in the database, e.g.:
```json
{"recovery_codes": [{"code": "lruvis83", "used_at": null}, ...]}
```

That doesn't feel very secure. Shouldn't they be hashed like passwords? If Kratos DB is leaked or accessed by an attacker, and they have passwords stolen from other websites, breaking the second factor becomes trivial.

**Comment by vinckr** [comment]
Hello @kszafran 
Apologies it has been some time since you asked.

I am not sure about all the implications, but the attack vector seems to be: 

- Attacker has access to the users password (plaintext) from somewhere
- Attacker also has access to a dump of the Kratos DB
- User is also re-using their password across services (more than 80% of users do this)


Is this a theoretical concern or part of your threat model? 
Can you share some more details? 
Usually if the attacker manages to dump your entire database, there are many different possible attacks as they are deep in your system already
I can imagine some use cases where this would be a realistic scenario. 

Due to the age of the discussion, I will mark it as answered for now, but feel free to come back to this and elaborate - feature requests are also welcome 🤗
